exchange/bitswap/network/ipfs_impl.go

```diff
@@ -38,15 +40,30 @@ type impl struct {
 }
 
 func (bsnet *impl) newStreamToPeer(ctx context.Context, p peer.ID) (inet.Stream, error) {
-
 	// first, make sure we're connected.
 	// if this fails, we cannot connect to given peer.
 	//TODO(jbenet) move this into host.NewStream?
 	if err := bsnet.host.Connect(ctx, peer.PeerInfo{ID: p}); err != nil {
 		return nil, err
 	}
 
-	return bsnet.host.NewStream(ctx, ProtocolBitswap, p)
+	s, err := bsnet.host.NewStream(ctx, ProtocolBitswap, p)
+	if err != nil {
+		return nil, err
+	}
+
+	// TEMP: trigger read handshake to detect protocl mismatch
+	// TODO: once the network has switched over entirely to the versioned
+	// protocol string, deprecate the old one
+	_, err = s.Read(nil)
+	if err != nil {
+		if strings.Contains(err.Error(), "protocol mismatch") {
+			return bsnet.host.NewStream(ctx, ProtocolBitswapOld, p)
+		}
+		return nil, err
+	}
```

Comment:
If we default for the ones with semver, then we only do the double RTT with old nodes and since our network is not that large, yet, we won't pay a big overhead. Each time we ship a new version of a wire protocol, we will pay this cost. Memorising the preferred multicodecs would be a big plus for this.